File: tests/0572_subtree_of_another_tree_test.py
import importlib
from algorithms.tree_node import TreeNode, make_tree

module = importlib.import_module("algorithms.0572_subtree_of_another_tree")


def test_is_subtree_same():
    t = make_tree([3, 4, 5, 1, 2])

    s = module.Solution()
    assert s.isSubtree(t[0], t[0])
    assert s.isSubtree(t[0], t[1])


def test_is_subtree1():
    a = make_tree([3, 4, 5, 1, 2])
    b = make_tree([4, 1, 2])

    s = module.Solution()
    assert s.isSubtree(a[0], b[0])


def test_is_subtree2():
    a = make_tree([1, 1])
    b = TreeNode(1)

    s = module.Solution()
    assert s.isSubtree(a[0], b)


def test_is_subtree3():
    a = make_tree([3, 4, 5, 1, 2, None, None, 0])
    b = make_tree([4, 1, 2])

    s = module.Solution()
    assert not s.isSubtree(a[0], b[0])


def test_is_subtree4():
    a = make_tree([3, 4, 5, 1, None, 2])
    b = make_tree([3, 1, 2])

    s = module.Solution()
    assert not s.isSubtree(a[0], b[0])


def test_is_subtree5():
    a = make_tree(
        [
            1,
            None,
            1,
            None,
            1,
            None,
            1,
            None,
            1,
            None,
            1,
            None,
            1,
            None,
            1,
            None,
            1,
            None,
            1,
            None,
            1,
            2,
        ]
    )
    b = make_tree([1, None, 1, None, 1, None, 1, None, 1, None, 1, 2])

    s = module.Solution()
    assert s.isSubtree(a[0], b[0])

File: tests/0572_subtree_of_another_tree_test.cpp
#include "0572_subtree_of_another_tree.cpp"
#include <doctest.h>
//#include <initializer_list>
#include <memory>

TEST_CASE("isSubTree - same tree")
{
    auto s = tree::make_tree({3, 4, 5, 1, 2});

    Solution sol = Solution();
    CHECK(sol.isSubtree(s[0].get(), s[0].get()));
    CHECK(sol.isSubtree(s[0].get(), s[1].get()));
}

TEST_CASE("isSubTree1")
{
    auto s = tree::make_tree({3, 4, 5, 1, 2});
    auto t = tree::make_tree({4, 1, 2});

    Solution sol = Solution();
    CHECK(sol.isSubtree(s[0].get(), t[0].get()));
}

TEST_CASE("isSubTree2")
{
    auto s0 = tree::make_tree({1, 1});
    auto t0 = std::make_unique<TreeNode>(1);

    Solution s = Solution();
    CHECK(s.isSubtree(s0[0].get(), t0.get()));
}

TEST_CASE("isSubTree3")
{
    auto s = tree::make_tree({3, 4, 5, 1, 2, std::nullopt, std::nullopt, 0});
    auto t = tree::make_tree({4, 1, 2});

    Solution sol = Solution();
    CHECK_FALSE(sol.isSubtree(s[0].get(), t[0].get()));
}

TEST_CASE("isSubTree4")
{
    auto s = tree::make_tree({3, 4, 5, 1, std::nullopt, 2});
    auto t = tree::make_tree({3, 1, 2});

    Solution sol = Solution();
    CHECK_FALSE(sol.isSubtree(s[0].get(), t[0].get()));
}

TEST_CASE("isSubTree5")
{
    auto s = tree::make_tree({1, std::nullopt,
                              1, std::nullopt,
                              1, std::nullopt,
                              1, std::nullopt,
                              1, std::nullopt,
                              1, std::nullopt,
                              1, std::nullopt,
                              1, std::nullopt,
                              1, std::nullopt,
                              1, std::nullopt,
                              1, 2});
    auto t =
        tree::make_tree({1, std::nullopt, 1, std::nullopt, 1, std::nullopt, 1, std::nullopt, 1, std::nullopt, 1, 2});

    Solution sol = Solution();
    CHECK(sol.isSubtree(s[0].get(), t[0].get()));
}
